### DISCUSSION TITLE
What "inference speed" is considered good? I have about 35-40.

### DISCUSSION THREAD

**Discussion by thinkloop** [blakeblackshear/frigate#3361] [Ask A Question] [answered]
![Screenshot from 2022-06-18 14-02-25@2x](https://user-images.githubusercontent.com/118814/174451052-9ee25cfa-e50d-4ae1-a095-681615595827.png)

**Comment by NickM-27** [comment, marked as answer]
Depends how many cameras you have. You want to be able to handle your detection speed (5 fps by default) so if your inference speed is 40 ms then that's 200 ms for 5 fps video. Mean you could theoretically handle 5 cameras each at 5 fps detection speed (real world may differ as some variables change this some)

**Reply by alez007** [reply]
this reply make no sense to me. Why would the first sentence imply the second one ? Why a 200ms for a 5 fps video implies that we could theoretically handle 5 cameral at 5 fps ?

**Reply by NickM-27** [reply]
200 ms for 5 fps video on one camera. There are 1000 ms in a second, so that means 200 ms for 1 camera = 1000 ms for 5 cameras. Worth noting though that as was explained below, the math is not quite that simple

**Reply by alez007** [reply]
ok thanks, got it, so what happens if we bypass a second ? frames start to get unprocessed ?

**Reply by NickM-27** [reply]
> ok thanks, got it, so what happens if we bypass a second ? frames start to get unprocessed ?

they are skipped and not processed, frigate values real time data over past data

**Comment by blakeblackshear** [comment]
Your camera fps is only loosely correlated. Frigate often runs multiple detections on a single frame, so you can easily see a higher detection fps than your camera fps. For a CPU, those times are not bad. A Coral usually has under 10ms inference times on decent hardware, so it's 3-4x faster and takes the load off the CPU. What CPU are you using?

**Reply by thinkloop** [reply]
I'm running a Intel(R) Core(TM) i9-9900K CPU @ 3.60GHz

**Reply by skynet01** [reply]
Unless you run Mac with M1 chip .. [then you will be in 10ms](https://github.com/blakeblackshear/frigate/discussions/1247) zone since they have Tensor processing built in.

**Comment by Nunu00** [comment]
Hi, I have installed frigate as an homeassistant addon on a raspberry pi 4 and since coral tpu is not currently available I am using the cpu as a detector. In the docs I had seen that the rpi4 had an inference speed of 10-15ms while I find myself with an inference speed of 350-400ms. Did I do something wrong or is there an error in the docs?

**Reply by NickM-27** [reply]
The inference speed you see in the docs is with a coral. And that's what you're aiming for, you're not doing anything wrong.

**Comment by alez007** [comment]
how does this inference speed scale with multiple detectors ? e.g. if I've got 6 detectors configured on a 6 core cpu and each detector tells me it's got an inference speed of 30 at 10fps... does this mean that each detector can support roughly 3 cameras until it starts skipping frames ? If so, why would frigate docs suggest having one detector per camera ?

**Reply by NickM-27** [reply]
> it's got an inference speed of 30 at 10fps... does this mean that each detector can support roughly 3 cameras until it starts skipping frame

if they all see activity at the same time I think it would skip frames at 3 cameras. especially because CPU detectors are more prone to slowing down when other things start taxing the CPU more.

we don't recommend CPU detectors at all, and the docs do not suggest having one coral / OpenVino / TensorRT per camera. 

the typical CPU detector is very slow so it makes sense that it would be needed.

**Reply by alez007** [reply]
I'm talking about this bit from the docs: https://docs.frigate.video/configuration/detectors/#cpu-detectors-not-recommended

```
When using CPU detectors, you can add a CPU detector per camera. Adding more detectors than the number of cameras should not improve performance.
```

One thing I don't understand is what do you mean by CPU detector ? is there any relation between cpu detectors and the actual cpu cores ? in the table, there's an inference speed allocated per detector... but there are 6 detectors... is it sensible to say that each of them has a 1000ms inference speed limit ?

I've got a decent proc, i9 9900 so I don't mind cpu detection, I just want to understand it, sorry if I'm being dumb.

Thanks

**Reply by NickM-27** [reply]
> I'm talking about this bit from the docs: https://docs.frigate.video/configuration/detectors/#cpu-detectors-not-recommended
> 
> ```
> When using CPU detectors, you can add a CPU detector per camera. Adding more detectors than the number of cameras should not improve performance.
> ```
> 
> One thing I don't understand is what do you mean by CPU detector ? 

a CPU detector is just the model being loaded in memory and being run on the CPU. It can have a number of threads: https://github.com/blakeblackshear/frigate/blob/master/frigate/edgetpu.py#L59

> is there any relation between cpu detectors and the actual cpu cores ? 

not directly I don't think but I wouldn't know for sure

>in the table, there's an inference speed allocated per detector... but there are 6 detectors... is it sensible to say that each of them has a 1000ms inference speed limit ?
> 
> Thanks

I'm not sure what table you are referring to, but . The CPU detector could take longer than 1000ms for an inference, there is no limit it just depends on the CPU. Of course there's 1000 ms in a second and anything past that frigate will skip frames so it continues processing real time data. 

> I've got a decent proc, i9 9900 so I don't mind cpu detection, I just want to understand it, sorry if I'm being dumb.

cpu usage isn't the only concern, it is also inference time. Also in 0.12 you can use your intel iGPU for running as a detector so that would be the recommendation instead of CPU detector

**Reply by alez007** [reply]
thanks! this is the table I'm speaking about, from frigate debug

![Screenshot 2023-01-20 at 15 33 03](https://user-images.githubusercontent.com/459478/213738400-ad6f9f45-eff4-4838-a26b-e7eea98f5f20.png)

```
Also in 0.12 you can use your intel iGPU for running as a detector so that would be the recommendation instead of CPU detector
```
I'd love to implement the above, will check how stable v0.12 is and its implementation 👍
